### PR TITLE
allow annotation on MachinePool to set externally managed

### LINKED ISSUE #2022: AWSMachinePool does not work in combination with cluster-autoscaler
/kind bug

**What steps did you take and what happened:**

- Create a workload cluster with the experimental EKS Control Plane
- Create a MachinePool without specifying replicas and create associated AWSMachinePool resources.
- Install cluster-autoscaler and configure it to autoscale the autoscaling group managed by the AWSMachinePool
- Create a deployment that will trigger cluster-autoscaler to scale up the autoscaling group.

This caused the AWSMachinePool controller and the cluster-autoscaler to both attempt to manage DesiredInstances. The result being the DesiredInstances value on the ASG was alternating between the value cluster-autoscaler computed as needed and the value of `replicas: 1` coming from the defaulted value on the MachinePool object.

**What did you expect to happen:**
cluster-autoscaler should manage the `DesiredInstances` of the autoscaling group and the AWSMachinePool controller should ignore changes to the `replicas` and `DesiredInstances` fields.

**Anything else you would like to add:**
Even though `replicas` is an optional field on the MachinePool resource, it defaults to 1, so leaving the value unspecified is insufficient to prevent the AWSMachinePool controller from attempting to managed the DesiredInstance count.

**Environment:**

- Cluster-api-provider-aws version: Commit: 3338cd4
- Kubernetes version: (use `kubectl version`): v.1.17.9
- OS (e.g. from `/etc/os-release`): Amazon Linux 2

### PR DESCRIPTION
**Pull request by mweibel**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
As a cluster operator, I can mark a MachinePool with the annotation `cluster.x-k8s.io/replicas-managed-by-autoscaler` to indicate it should not enforce the `MachinePool.replicas` field but instead reconcile it to match the ASGs desiredCapacity field. This allows me to use an external autoscaler which does not come into conflict with the MachinePool reconciliation loop. 

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #2022

**Special notes for your reviewer**:
The annotation is the same as in kubernetes-sigs/cluster-api-provider-azure#2444.

Unfortunately, I did not yet manage to make the added test in `awsmachinepool_controller_test.go` pass. This is also why the PR is marked as WIP/Draft. The test fails with:

```
I0822 10:22:31.013336   20405 awsmachinepool_controller.go:256]  "msg"="Setting MachinePool replicas to ASG DesiredCapacity" "external"=1 "local"=0
    awsmachinepool_controller_test.go:313: 
        Expected
            <errors.aggregate | len:1, cap:1>: [
                <*meta.NoKindMatchError | 0x14000263ec0>{
                    GroupKind: {
                        Group: "cluster.x-k8s.io",
                        Kind: "MachinePool",
                    },
                    SearchedVersions: ["v1beta1"],
                },
            ]
        to be nil
```

I believe the issue is that the API client does not know about the MachinePool CRD. I added a fake machinepool CRD just like the one for Machines/Clusters but it did not resolve the issue. Would be great if somebody could have a look at it. 

**Checklist**:

- [x] squashed commits
- [ ] includes documentation
- [x] adds unit tests
- [ ] adds or updates e2e tests

### CODE REVIEW

**Comment by sedefsavas** [comment]
There is a related PR to this in CAPI:
https://github.com/kubernetes-sigs/cluster-api/pull/7107

**Comment by mweibel** [comment]
@sedefsavas yes saw that one, thanks. Once that is merged and released, this code could reuse the const from CAPI. 

I still didn't figure out the issue with the failing test. @richardcase had a look at it IIRC but not sure what the state is.

**Comment by mweibel** [comment]
I'm trying every now and then when I find some time to figure out the reason why the MachinePool Kind is not found. I'd love some help on this topic :)

**Comment by richardcase** [comment]
I had a quick look at this and i also couldn't work out why it was failing.

File: controllers/suite_test.go
Comment (Skarlso) [inline controllers/suite_test.go:50]: Shouldn't this be `scheme.Scheme`? Maybe that's what's causing issues.

File: controllers/suite_test.go
Comment (Skarlso) [inline controllers/suite_test.go:51]: Or this one be also `scheme`. :D I'm actually unsure. :D

File: controllers/suite_test.go
Comment (Skarlso) [inline controllers/suite_test.go:50]: Yep, it's this. This is actually a compile time failure for me if I check this out.

Comment (richardcase) [reply controllers/suite_test.go:50]: I have it as `scheme.Scheme` and i get the `meta.NoKindMatchError` error. I'm stumped

**Comment by richardcase** [comment]
@mweibel - i've got to the bottom of the `NoKindMatchError` error you are seeing. I am going to make a small PR later today. Once that merges you will need to rebase.

**Comment by richardcase** [comment]
That change has merged. So if you rebase you should be able to get past the `NoKindMatchError` error.

**Comment by mweibel** [comment]
@richardcase oh - that change makes sense :+1: thanks for investigating and fixing it! Will rebase as soon as I'm able (currently on vacation)

File: controllers/suite_test.go
Comment (mweibel) [reply controllers/suite_test.go:50]: possibly - sorry, my second commit was actually just WIP garbage commit where I tried to find the cause. Will revert those unnecessary changes.

File: pkg/cloud/scope/machinepool.go
Comment (richardcase) [inline pkg/cloud/scope/machinepool.go:53]: Ideally we'd want to not export this and provide a `PatchMachinePool` function on the scope. But i'm not so concerned about this

**Comment by richardcase** [comment]
e2e tests are passing :+1:

**Comment by Skarlso** [comment]
Ah, that's a good catch. We actually ran into a similar problem with one of the issues. That was patching some template which ended up being out of sync with what you basically had locally. So once you applied again, the changes were lost.

Interesting thought. I don't see an option out of that other than having an unmanaged field perhaps.

**Comment by mweibel** [comment]
@richardcase that's a good catch indeed. We're using ArgoCD where you can ignore differences. A user with a GitOps solution would need to do that:
```yaml
  ignoreDifferences:
    - group: cluster.x-k8s.io
      kind: MachinePool
      jsonPointers:
        - /spec/replicas
```

I agree that this is not beautiful. One way this could be different when CAPI would provide e.g. a `status.replicas` field which overrules `spec.replicas` field in case it's set, so `spec.replicas` would only be considered for the first reconciliation. Not sure if that's a good approach though.

File: pkg/cloud/scope/machinepool.go
Comment (mweibel) [reply pkg/cloud/scope/machinepool.go:53]: I wondered if I should do it that way or provide a specific function on the scope. I don't mind changing this.

**Comment by dthorsen** [comment]
An even better option is to completely omit the replicas field from your manifest.

**Comment by richardcase** [comment]
> An even better option is to completely omit the replicas field from your manifest.

@dthorsen - i think thats probably the best solution. 

What do you think to the following:

- Change the MachinePool update logic that in the `AWSMachinePool` reconciler so that it only updates the replicas if its not nil
- Add a documentation for externally managed replicas that includes:
  - The option/recommendation of not setting replicas on MachinePool when using externally managed
  - The sample above for ignoring the field with argo

**Comment by Skarlso** [comment]
> An even better option is to completely omit the replicas field from your manifest.

Wouldn't that still fail? Because it's nil locally, but set upstream?

**Comment by richardcase** [comment]
> Wouldn't that still fail? Because it's nil locally, but set upstream?

Its omit it from the upstream `MachinePool`

File: pkg/cloud/scope/machinepool.go
Comment (AverageMarcus) [inline pkg/cloud/scope/machinepool.go:46]: Am I correct that we'd import this value from CAPI once [this PR](https://github.com/kubernetes-sigs/cluster-api/pull/7107) is merged rather than hardcoding it in here?

**Comment by mweibel** [comment]
> > Wouldn't that still fail? Because it's nil locally, but set upstream?
> 
> Its omit it from the upstream `MachinePool`

@richardcase looking at the webhook code, I believe this wouldn't work since the webhook would set it anyway:
https://github.com/kubernetes-sigs/cluster-api/blob/main/exp/api/v1beta1/machinepool_webhook.go#L54..L56

or do I miss something there?

File: pkg/cloud/scope/machinepool.go
Comment (mweibel) [reply pkg/cloud/scope/machinepool.go:46]: yes, I think that would make sense.

File: pkg/cloud/scope/machinepool.go
Comment (mweibel) [reply pkg/cloud/scope/machinepool.go]: adjusted :+1:

**Comment by mweibel** [comment]
I adjusted the patch helper so it's a function on the MachinePool scope instead of exposing the patch helper itself. 

@richardcase how to handle MachinePool.replicas: I don't believe there is currently a better option. I added some documentation which might help in this case. 

Would be great to have another review on this :+1:

File: docs/book/src/topics/machinepools.md
Comment (AverageMarcus) [inline docs/book/src/topics/machinepools.md:112]:
```suggestion
Two providers are possible to use with CAPA MachinePools: `clusterapi`, or `aws`.
```

**Review by AverageMarcus — CHANGES_REQUESTED** [review]

File: docs/book/src/topics/machinepools.md
Comment (Ankitasw) [reply docs/book/src/topics/machinepools.md:112]: Good catch 😄

Comment (mweibel) [reply docs/book/src/topics/machinepools.md:112]: oh 😄 sorry about that one! good catch indeed.

**Comment by mweibel** [comment]
docs typo is fixed 👍

**Review by AverageMarcus — APPROVED** [review]

**Comment by mweibel** [comment]
FTR this PR has been updated with the updated annotation key/value from kubernetes-sigs/cluster-api#7107.

**Pull request merged**
